cloudbuild.yaml

```diff
@@ -49,20 +50,46 @@ steps:
       echo "Get cloud build project configuration values"
       source .env.cloud_build 
 
+      echo "Copy test coverage report from docker volume to a google cloud storage bucket."
       timestamp=$(date +%s)
+      gsutil cp ./coverage/coverage.json "gs://${TEST_COVERAGE_BUCKET_NAME}/${BRANCH_NAME}/${timestamp}-${COMMIT_SHA}.json"
 
-      echo "Copy test coverage report from docker compose volume to Google Gloud Storage."
-      gsutil cp ./coverage/coverage.json "gs://hopic-test-coverage-reports/${BRANCH_NAME}/$timestamp-${COMMIT_SHA}.json"
-
-  - id: "Tear down tests"
-    name: "gcr.io/cloud-builders/docker"
-    script: |
-      #!/usr/bin/env bash
-      set -o errexit
-      set -o pipefail
-      set -o nounset
+      # Install jq and bc 
+      apt-get update && apt-get install -y jq bc
+      
+      # Compare total coverage to failure threshold.
+      total_coverage=$(jq '.totals.percent_covered' ./coverage/coverage.json | awk '{printf "%.2f\n", $1}')
+      if (( $(echo "${total_coverage} < ${TEST_COVERAGE_THRESHOLD}" | bc -l) )); then
+        echo "Coverage (${total_coverage}%) is below the threshold of (${TEST_COVERAGE_THRESHOLD}%) - please add more unit tests!"
+        echo "For more details, read the test coverage report logged in 'Run tests' cloud build step."
+        # Fail build if main branch.
+        if [[ "${BRANCH_NAME}" == 'main' ]]; then  
+          echo "Failing build for insufficent test coverage."  
+          exit 1
```

Comment:
Instead of exiting immediately, I think it'd be better if exiting happened after all of the stats had been calculated and logged (coverage threshold _and_ deltas). That way the information to be found in the build output is consistent each time.

Since I think we might want additional failure states, which will not be mutually exclusive, I'd suggest tracking failure reasons in an array. Declare an empty bash array higher up with `failure_reasons=()` and append failure causes to it with `failure_reasons+=("Another reason this failed")`. You can check whether the array is empty with `if [ ${#failure_reasons[@]} -eq 0 ]; ...` at the end. If it's not empty, echo any items in it and then exit with a failure state.

Reply:
Implemented this - logging first, and not failing on main branch  (only if not-main branch (but also commented that out as you're right we may not want to have this fail.))